### PR TITLE
CONSOLE-3916: Direct OIDC tech debt

### PR DESCRIPTION
**Pull request by stlaz**
Adds unit tests.
Fixes a bug in the AsyncCache where it would never return an error on init.

/assign @jhadvig

### CODE REVIEW

**Comment by stlaz** [comment]
/hold
I've found another bug with how the async cache handles contexts

File: pkg/auth/auth_oidc_test.go
Comment (stlaz) [inline pkg/auth/auth_oidc_test.go:279]: The truth appears to be a bit more complicated. Token verification actually wouldn't use the context that's pushed to `provider.Verify()` to update the keys, but it uses the context it was constructed with.

However, the async cache implementation we've got will have long cancelled that context for the first item retrieved during initialization thanks to the `PollUntilContextTimeout` logic used there.

**Comment by stlaz** [comment]
/hold cancel
another asynccache bug fixed in the last commit

File: pkg/auth/sessions/combined_sessions.go
Comment (jhadvig) [inline pkg/auth/sessions/combined_sessions.go:176]: so previously we have been waiting for the token to expire ?

**Review by jhadvig — APPROVED** [review]
/lgtm
/approve

**Comment by stlaz** [comment]
/hold
for https://github.com/openshift/console/pull/13703

**Comment by stlaz** [comment]
/hold cancel
@jhadvig if I may ask you to retag

File: pkg/auth/sessions/combined_sessions.go
Comment (stlaz) [reply pkg/auth/sessions/combined_sessions.go:176]: Yes. This is still not great since, as I've noticed, the session deletion code will likely never be called.
A fix should come in a different PR.

**Comment by jhadvig** [comment]
QE Approver: 
/assign @yapei
Docs Approver:
/assign @opayne1
PX Approver:
/assign @RickJWagner

**Comment by jhadvig** [comment]
@yapei here we will need to do just regression testing, no additional TCs are needed, since the change is mainly adding unit tests and refactoring code.

**Comment by yanpzhan** [comment]
Normal regression test passed.
/label qe-approved

**Pull request merged**
